src/interop/js-interop/js-types.md

```diff
@@ -0,0 +1,242 @@
+| `JSNumber`, `JSBoolean`, `JSString` | `num`, `int`, `double`, `bool`, `String` |
+| `JSExportedFunction`                | `Function`                               |
+| `JSArray<T extends JSAny?>`         | `List<T extends JSAny?>`                 |
+| `JSPromise<T extends JSAny?>`       | `Future<T extends JSAny?>`               |
+| Typed arrays like `JSUint8Array`    | `dart:typed_data`                        |
+| `JSBoxedDartObject`                 | Opaque Dart value                        |
+{:.table .table-striped}
+</div>
+
+{{site.alert.warn}}
+Conversions might have different costs depending on the compiler, so prefer to
+only convert values if you need to. Conversions also may or may not produce a
+new value. This doesn’t matter for immutable values like numbers, but does
+matter for types like `List`. A conversion to a `JSArray` may produce a new
+value by copying or may not, so do not rely on modifications to the `JSArray` to
+affect the `List`. Typed array conversions have a similar limitation. Look up
+the specific conversion function for more details.
+{{site.alert.end}}
+
+## Requirements on `external` declarations and `Function.toJS`
+
+In order to ensure type safety and consistency, the compiler places requirements
+on what types can flow into and out of JS. Passing arbitrary Dart values into JS
+is not allowed. Instead, the compiler requires users to use a compatible interop
+type like a JS type or a primitive, which would then be implicitly converted by
+the compiler. For example, these would be allowed:
+
+{:.good}
+```dart
+@JS()
+external void primitives(String a, int b, double c, num d, bool e);
+```
+
+{:.good}
+```dart
+@JS()
+external JSArray jsTypes(JSObject _, JSString __);
+```
+
+{:.good}
+```dart
+extension type InteropType(JSObject _) {}
+
+@JS()
+external InteropType get interopType;
+```
+
+Whereas these would return an error:
+
+{:.bad}
+```dart
+@JS()
+external Function get function;
+```
+
+{:.bad}
+```dart
+@JS()
+external set list(List _);
+```
+
+These same requirements exist when you use [`Function.toJS`] to make a Dart
+function callable in JS. The values that flow into and out of this callback must
+be a compatible interop type or a primitive.
+
+If you use a Dart primitive like `String`, an implicit conversion happens in the
+compiler to convert that value from a JS value to a Dart value. If performance
+is critical and you don’t need to examine the contents of the string, then using
+`JSString` instead to avoid the conversion cost might make sense like in the
+second example.
+
+## Compatibility, type checks, and casts
+
+The representation type of JS types may differ based on the compiler. This
+affects runtime type-checking and casts. Therefore, almost always avoid `is`
+checks where the value is an interop type or where the target type is an interop
+type:
+
+{:.bad}
+```dart
+void f(JSAny a) {
+  if (a is String) { … }
+}
+```
+{:.bad}
+```dart
+void f(JSObject o) {
```

Comment:
did you mean to use JSAny here (otherwise it's the same type as the `is` check)

Reply:
Yes, good catch!